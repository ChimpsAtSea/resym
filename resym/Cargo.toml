[package]
name = "resym"
version.workspace = true
edition = "2021"
rust-version = "1.67"
build = "build.rs"

[package.metadata.winres]
ProductName = "resym"
ProductVersion = "0.3.0"
FileDescription = "Cross-platform tool that allows browsing and extracting C and C++ type declarations from PDB files."
LegalCopyright = "Copyright Â© resym developers"

[dependencies]
resym_core = { version = "0.3", path = "../resym_core" }

eframe = { version = "0.21", features = ["persistence"] }
tinyfiledialogs = "3.9"
serde = "1.0"
syntect = "5.0"
anyhow = "1.0"
log = "0.4"
memory_logger = { version = "0.1", features = ["blocking"] }
crossbeam-channel = "0.5"

[target.'cfg(windows)'.build-dependencies]
winres = "0.1"
image = "0.24"
