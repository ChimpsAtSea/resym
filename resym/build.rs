#[cfg(windows)]
fn main() {
    let target = std::env::var("TARGET").unwrap();
    if target.contains("wasm") {
        return;
    }

    // Convert ICO to raw pixels at build time. This avoids adding the `image` crate in `resym`
    // only to display an icon.
    let icon_bytes_path = std::path::Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/resources/resym_96.bin"
    ));
    // Only generate the binary file if it doesn't exist
    if !icon_bytes_path.exists() {
        let image_path = concat!(env!("CARGO_MANIFEST_DIR"), "/resources/resym_96.ico");
        let image = image::open(image_path)
            .expect("failed to open 96x96 ICO")
            .into_rgba8();
        let image_bytes = image.into_raw();

        std::fs::write(icon_bytes_path, image_bytes).expect("failed to save 96x96 icon bytes");
    }

    // Compile the resource file with metadata and an icon
    let mut res = winres::WindowsResource::new();
    res.set_icon("resources/resym_256.ico");
    res.compile().expect("Windows resource compilation failed");
}

#[cfg(unix)]
fn main() {}
