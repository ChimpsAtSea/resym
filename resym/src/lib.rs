#[cfg(target_arch = "wasm32")]
mod frontend;
#[cfg(target_arch = "wasm32")]
mod mode;
#[cfg(target_arch = "wasm32")]
mod module_tree;
#[cfg(target_arch = "wasm32")]
mod module_tree_view;
#[cfg(target_arch = "wasm32")]
mod resym_app;
#[cfg(target_arch = "wasm32")]
mod settings;
#[cfg(target_arch = "wasm32")]
mod syntax_highlighting;
#[cfg(target_arch = "wasm32")]
mod ui_components;

#[cfg(target_arch = "wasm32")]
use memory_logger::blocking::MemoryLogger;

#[cfg(target_arch = "wasm32")]
use resym_app::ResymApp;

#[cfg(target_arch = "wasm32")]
use eframe::wasm_bindgen::{self, prelude::*};

#[cfg(target_arch = "wasm32")]
#[wasm_bindgen]
pub struct WebHandle {
    runner: eframe::WebRunner,
}

#[cfg(target_arch = "wasm32")]
#[wasm_bindgen]
impl WebHandle {
    /// Installs a panic hook, then returns.
    #[allow(clippy::new_without_default)]
    #[wasm_bindgen(constructor)]
    pub fn new() -> Self {
        Self {
            runner: eframe::WebRunner::new(),
        }
    }

    /// Call this once from JavaScript to start your app.
    #[wasm_bindgen]
    pub async fn start(&self, canvas_id: &str) -> Result<(), wasm_bindgen::JsValue> {
        let logger = MemoryLogger::setup(log::Level::Info).expect("application creation");

        self.runner
            .start(
                canvas_id,
                eframe::WebOptions::default(),
                Box::new(|cc| Box::new(ResymApp::new(cc, logger).expect("application creation"))),
            )
            .await
    }

    #[wasm_bindgen]
    pub fn destroy(&self) {
        self.runner.destroy();
    }

    /// The JavaScript can check whether or not your app has crashed:
    #[wasm_bindgen]
    pub fn has_panicked(&self) -> bool {
        self.runner.has_panicked()
    }

    #[wasm_bindgen]
    pub fn panic_message(&self) -> Option<String> {
        self.runner.panic_summary().map(|s| s.message())
    }

    #[wasm_bindgen]
    pub fn panic_callstack(&self) -> Option<String> {
        self.runner.panic_summary().map(|s| s.callstack())
    }
}
