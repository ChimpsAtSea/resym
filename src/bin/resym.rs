use anyhow::{anyhow, Result};
use crossbeam_channel::{Receiver, Sender};
use eframe::{
    egui::{self, Visuals},
    epi,
};
use egui::{ScrollArea, TextStyle};
use memory_logger::blocking::MemoryLogger;
use serde::{Deserialize, Serialize};
use tinyfiledialogs::open_file_dialog;

use std::sync::{Arc, RwLock};

use resym::{
    backend::{Backend, BackendCommand},
    frontend::{FrontendCommand, FrontendController},
    PKG_NAME, PKG_VERSION,
};

fn main() -> Result<()> {
    let logger = MemoryLogger::setup(log::Level::Info)?;
    let app = ResymApp::new(logger)?;
    let native_options = eframe::NativeOptions::default();
    eframe::run_native(Box::new(app), native_options);
}

#[derive(Serialize, Deserialize)]
struct ResymAppSettings {
    use_light_theme: bool,
    search_case_insensitive: bool,
    search_use_regex: bool,
    print_header: bool,
    reconstruct_dependencies: bool,
    print_access_specifiers: bool,
}

impl Default for ResymAppSettings {
    fn default() -> Self {
        Self {
            use_light_theme: false,
            search_case_insensitive: true,
            search_use_regex: false,
            print_header: true,
            reconstruct_dependencies: true,
            print_access_specifiers: true,
        }
    }
}

struct EguiFrontendController {
    tx_ui: Sender<FrontendCommand>,
    ui_frame: RwLock<Option<epi::Frame>>,
}

impl EguiFrontendController {
    pub fn new(tx_ui: Sender<FrontendCommand>) -> Self {
        Self {
            tx_ui,
            ui_frame: RwLock::new(None),
        }
    }

    fn set_ui_frame(&self, ui_frame: epi::Frame) -> Result<()> {
        match self.ui_frame.write() {
            Err(_) => Err(anyhow!("Failed to update `ui_frame`")),
            Ok(mut ui_frame_opt) => {
                *ui_frame_opt = Some(ui_frame);
                Ok(())
            }
        }
    }
}

impl FrontendController for EguiFrontendController {
    /// Used by the backend to send us commands and trigger a UI update
    fn send_command(&self, command: FrontendCommand) -> Result<()> {
        self.tx_ui.send(command)?;
        // Force the UI backend to call our app's update function on the other end
        if let Ok(ui_frame_opt) = self.ui_frame.try_read() {
            if let Some(ui_frame) = ui_frame_opt.as_ref() {
                ui_frame.request_repaint();
            }
        }
        Ok(())
    }
}

struct ResymApp {
    logger: &'static MemoryLogger,
    filtered_type_list: Vec<(String, pdb::TypeIndex)>,
    selected_row: usize,
    search_filter: String,
    reconstructed_type_content: String,
    console_content: Vec<String>,
    settings_wnd_open: bool,
    settings: ResymAppSettings,
    rx_ui: Receiver<FrontendCommand>,
    frontend_controller: Arc<EguiFrontendController>,
    backend: Backend,
}

impl<'p> ResymApp {
    fn new(logger: &'static MemoryLogger) -> Result<Self> {
        let (tx_ui, rx_ui) = crossbeam_channel::unbounded::<FrontendCommand>();
        let frontend_controller = Arc::new(EguiFrontendController::new(tx_ui));
        let backend = Backend::new(frontend_controller.clone())?;

        Ok(Self {
            logger,
            filtered_type_list: vec![],
            selected_row: usize::MAX,
            search_filter: String::default(),
            reconstructed_type_content: String::default(),
            console_content: vec![],
            settings_wnd_open: false,
            settings: ResymAppSettings::default(),
            rx_ui,
            frontend_controller,
            backend,
        })
    }

    fn process_ui_commands(&mut self) {
        while let Ok(cmd) = self.rx_ui.try_recv() {
            match cmd {
                FrontendCommand::UpdateReconstructedType(data) => {
                    self.reconstructed_type_content = data;
                }

                FrontendCommand::UpdateFilteredSymbols(filtered_symbols) => {
                    self.filtered_type_list = filtered_symbols;
                    self.selected_row = usize::MAX;
                }
            }
        }
    }

    fn draw_menu_bar(&mut self, ui: &mut egui::Ui, frame: &epi::Frame) {
        egui::menu::bar(ui, |ui| {
            ui.menu_button("File", |ui| {
                if ui.button("Open PDB file").clicked() {
                    if let Some(file_path) = open_file_dialog(
                        "Select a PDB file",
                        "",
                        Some((&["*.pdb"], "PDB files (*.pdb)")),
                    ) {
                        if let Err(err) = self
                            .backend
                            .send_command(BackendCommand::LoadPDB(file_path))
                        {
                            log::error!("Failed to load the PDB file: {}", err);
                        } else {
                            let result =
                                self.backend
                                    .send_command(BackendCommand::UpdateSymbolFilter(
                                        String::default(),
                                        false,
                                        false,
                                    ));
                            if let Err(err) = result {
                                log::error!("Failed to update type filter value: {}", err);
                            }
                        }
                    }
                }
                if ui.button("Settings").clicked() {
                    self.settings_wnd_open = true;
                }
                if ui.button("Exit").clicked() {
                    frame.quit();
                }
            });
        });
    }

    fn draw_symbol_list(&mut self, ui: &mut egui::Ui) {
        let num_rows = self.filtered_type_list.len();
        const TEXT_STYLE: TextStyle = TextStyle::Body;
        let row_height = ui.text_style_height(&TEXT_STYLE);
        ui.with_layout(
            egui::Layout::top_down(egui::Align::Min).with_cross_justify(true),
            |ui| {
                ScrollArea::vertical()
                    .auto_shrink([false, false])
                    .show_rows(ui, row_height, num_rows, |ui, row_range| {
                        for row_index in row_range {
                            let (symbol_name, type_index) = &self.filtered_type_list[row_index];

                            if ui
                                .selectable_label(self.selected_row == row_index, symbol_name)
                                .clicked()
                            {
                                self.selected_row = row_index;
                                let result =
                                    self.backend.send_command(BackendCommand::ReconstructType(
                                        *type_index,
                                        self.settings.print_header,
                                        self.settings.reconstruct_dependencies,
                                        self.settings.print_access_specifiers,
                                    ));
                                if let Err(err) = result {
                                    log::error!("Failed to reconstruct type: {}", err);
                                }
                            }
                        }
                    });
            },
        );
    }

    fn draw_console(&mut self, ui: &mut egui::Ui) {
        // Update console
        self.console_content
            .extend(self.logger.read().lines().map(|s| s.to_string()));
        self.logger.clear();

        const TEXT_STYLE: TextStyle = TextStyle::Monospace;
        let row_height = ui.text_style_height(&TEXT_STYLE);
        let num_rows = self.console_content.len();
        ScrollArea::vertical()
            .auto_shrink([false, false])
            .stick_to_bottom()
            .show_rows(ui, row_height, num_rows, |ui, row_range| {
                for row_index in row_range {
                    ui.add(
                        egui::TextEdit::singleline(&mut self.console_content[row_index].as_str())
                            .font(egui::TextStyle::Monospace)
                            .desired_width(f32::INFINITY),
                    );
                }
            });
    }

    fn draw_settings_window(&mut self, ctx: &egui::Context) {
        egui::Window::new("Settings")
            .anchor(egui::Align2::CENTER_CENTER, [0.0; 2])
            .open(&mut self.settings_wnd_open)
            .auto_sized()
            .collapsible(false)
            .show(ctx, |ui| {
                ui.label("Theme");
                ui.checkbox(&mut self.settings.use_light_theme, "Use light theme");
                ui.add_space(5.0);

                ui.label("Search");
                ui.checkbox(
                    &mut self.settings.search_case_insensitive,
                    "Case insensitive",
                );
                ui.checkbox(
                    &mut self.settings.search_use_regex,
                    "Enable regular expressions",
                );
                ui.add_space(5.0);

                ui.label("Type reconstruction");
                ui.checkbox(&mut self.settings.print_header, "Print header");
                ui.checkbox(
                    &mut self.settings.reconstruct_dependencies,
                    "Print definitions of referenced types",
                );
                ui.checkbox(
                    &mut self.settings.print_access_specifiers,
                    "Print access specifiers",
                );
            });
    }
}

impl epi::App for ResymApp {
    fn name(&self) -> &str {
        PKG_NAME
    }

    /// Called once before the first frame.
    fn setup(
        &mut self,
        _ctx: &egui::Context,
        frame: &epi::Frame,
        storage: Option<&dyn epi::Storage>,
    ) {
        log::info!("{} {}", PKG_NAME, PKG_VERSION);
        // If this fails, let it burn
        self.frontend_controller
            .set_ui_frame(frame.clone())
            .unwrap();

        // Load settings on launch
        if let Some(storage) = storage {
            self.settings = epi::get_value(storage, epi::APP_KEY).unwrap_or_default()
        }
    }

    fn save(&mut self, storage: &mut dyn epi::Storage) {
        // Save settings on shutdown
        epi::set_value(storage, epi::APP_KEY, &self.settings);
    }

    /// Called each time the UI needs repainting, which may be many times per second.
    /// Put your widgets into a `SidePanel`, `TopPanel`, `CentralPanel`, `Window` or `Area`.
    fn update(&mut self, ctx: &egui::Context, frame: &epi::Frame) {
        // Process incoming commands, if any
        self.process_ui_commands();

        // Update theme
        let theme = if self.settings.use_light_theme {
            Visuals::light()
        } else {
            Visuals::dark()
        };
        ctx.set_visuals(theme);

        // Draw "Settings" window if open
        self.draw_settings_window(ctx);

        egui::TopBottomPanel::top("top_panel").show(ctx, |ui| {
            // The top panel is often a good place for a menu bar
            self.draw_menu_bar(ui, frame);
        });

        egui::SidePanel::left("side_panel")
            .default_width(250.0)
            .width_range(100.0..=f32::INFINITY)
            .show(ctx, |ui| {
                ui.label("Search");
                ui.add_space(4.0);

                if ui.text_edit_singleline(&mut self.search_filter).changed() {
                    // Update filtered list if filter has changed
                    let result = self
                        .backend
                        .send_command(BackendCommand::UpdateSymbolFilter(
                            self.search_filter.clone(),
                            self.settings.search_case_insensitive,
                            self.settings.search_use_regex,
                        ));
                    if let Err(err) = result {
                        log::error!("Failed to update type filter value: {}", err);
                    }
                }
                ui.add_space(4.0);

                // Display list of symbol names
                self.draw_symbol_list(ui);
            });

        // Bottom panel containing the console
        egui::TopBottomPanel::bottom("bottom_panel")
            .default_height(100.0)
            .show(ctx, |ui| {
                // Console panel
                ui.vertical(|ui| {
                    ui.label("Console");
                    ui.add_space(4.0);

                    self.draw_console(ui);
                });
            });

        egui::CentralPanel::default().show(ctx, |ui| {
            // The central panel the region left after adding TopPanel's and SidePanel's
            ui.label("Reconstructed type(s) - C++");
            ui.add_space(4.0);

            // Symbol dump area
            egui::ScrollArea::vertical()
                .auto_shrink([false, false])
                .show(ui, |ui| {
                    ui.add(
                        egui::TextEdit::multiline(&mut self.reconstructed_type_content.as_str())
                            .code_editor()
                            .desired_width(f32::INFINITY),
                    );
                });
        });
    }
}
