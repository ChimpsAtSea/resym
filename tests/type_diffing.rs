use std::path::Path;

use resym::{diffing::diff_type_by_name, pdb_file::PdbFile};

const TEST_PDB_FROM_FILE_PATH: &str = "tests/data/test_diff_from.pdb";
const TEST_PDB_TO_FILE_PATH: &str = "tests/data/test_diff_to.pdb";
const TEST_CASES: &[&str] = &[
    "UserStructAddAndReplace",
    "UserStructRemove",
    "UserStructAdd",
    "RemovedStruct",
    "NewStruct",
    "TypeNotFound",
];

#[test]
fn test_struct_diffing_no_dependencies_without_line_numbers() {
    let pdb_file_from = PdbFile::load_from_file(Path::new(TEST_PDB_FROM_FILE_PATH))
        .expect("load test_diff_from.pdb");
    let pdb_file_to =
        PdbFile::load_from_file(Path::new(TEST_PDB_TO_FILE_PATH)).expect("load test_diff_to.pdb");

    for test_case_type_name in TEST_CASES {
        let diffed_type = diff_type_by_name(
            &pdb_file_from,
            &pdb_file_to,
            test_case_type_name,
            false,
            false,
            false,
            false,
        );
        insta::assert_snapshot!(diffed_type);
    }
}

#[test]
fn test_struct_diffing_no_dependencies_with_line_numbers() {
    let pdb_file_from = PdbFile::load_from_file(Path::new(TEST_PDB_FROM_FILE_PATH))
        .expect("load test_diff_from.pdb");
    let pdb_file_to =
        PdbFile::load_from_file(Path::new(TEST_PDB_TO_FILE_PATH)).expect("load test_diff_to.pdb");

    for test_case_type_name in TEST_CASES {
        let diffed_type = diff_type_by_name(
            &pdb_file_from,
            &pdb_file_to,
            test_case_type_name,
            false,
            false,
            false,
            true,
        );
        insta::assert_snapshot!(diffed_type);
    }
}
