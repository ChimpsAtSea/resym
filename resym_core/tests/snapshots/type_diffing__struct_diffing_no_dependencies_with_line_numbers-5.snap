---
source: tests/type_diffing.rs
expression: diffed_type
---
  1 |+
  2 |+struct NewStruct { // Size=0x4
  3 |+  /* 0x0000 */ int32_t field;
  4 |+};

